lib/src/source/hosted.dart

```diff
@@ -227,14 +227,16 @@ class HostedSource extends Source {
         version == null || version >= _minVersionForShorterHostedSyntax;
 
     if (description is String) {
-      if (!canUseShorthandSyntax) {
-        throw FormatException('The syntax `hosted: <hosted-url> requires a` '
-            'minimum Dart SDK constraint of $_minVersionForShorterHostedSyntax!');
+      // Old versions of pub (pre Dart 2.15) interpret `hosted: foo` as
+      // `hosted: {name: foo, url: <default>}`.
+      // For later versions, we treat it as `hosted: {name: <inferred>,
+      // url: foo}` if a user opts in by raising their min SDK environment.
+      if (canUseShorthandSyntax) {
+        return HostedDescription(
+            packageName, validateAndNormalizeHostedUrl(description));
+      } else {
+        return HostedDescription(description, defaultUrl);
```

Comment:
Hmm, I know I said I thought this was smart... but thinking about it now, it has a clear downside:
If you write `hosted: <url>` with the wrong SDK constraint, then you'll get a weird error about the name containing illegal characters or worse some package can't be found :see_no_evil: 
You won't be told that the format `hosted: <url>` is only supported with lower bound SDK constraint of 2.15. Hence, users will try all sorts of weird things, instead of making the right choice and increasing the lower bound SDK constraint.

---------------------

This makes a lot of sense when interpreting the `pubspec.yaml` of a package that we depend on, though, because here we have to be compatible.

But if we see `hosted: foo` in the root `pubspec.yaml` we could choose to complain that lower-bound SDK constraint isn't 2.15. This is a breaking change, but it's only breaking the root package `pubspec.yaml` -- this is usually a file the user can change, it's not breaking any packages that have been published.

---------------

Another option is that we could choose to throw a `FormatException` if:
 * Lower bound SDK constraint is less than 2.15 (short syntax not supported), and,
 * We see `hosted: <value>`, and,
 * `<value>` is not a valid package name (`^(?:[a-zA-Z_]+)(\.?:[a-zA-Z_]+)*$`, see https://github.com/dart-lang/pub/blob/master/lib/src/pubspec.dart#L26-L32)

This way we get to point out the fact that `hosted: <url>` doesn't work pre-2.15 SDK constraint, and `hosted: <value>` is not a valid package name.

It's not perfect because `hosted: pub.example.com` would still cause some obscure error about `package pub.example.com not found on pub.dev`...

--------

Any ideas?

Reply:
Distinguishing between the root pubspec and pubspecs of dependencies will further complicate parsing (we'd have to add this option to `Pubspec.fromMap` and pass it down to the sources). I think I like the option of throwing a `FormatException` for values that aren't valid package names for Dart <2.15.

> It's not perfect because hosted: `pub.example.com` would still cause some obscure error about `package pub.example.com not found on pub.dev` ...

We require hosted urls to have a scheme (regardless of the syntax used), so at least there's no ambiguity for valid hosts...

From the comment on the regex, it looks like package names with dots in them are a Google-internal thing not realistically used by anyone else. So maybe we could just consider package names with a single segment (`^[a-zA-Z_]+$`) when attempting to check whether a url was meant. Then this would only be a problem if some pubspec in Google-internal code used `hosted: foo.bar.baz` which I imagine is somewhat easier (for you :D) to find and migrate?

Alternatively, we could also emit a warning only instead of refusing to accept the pubspec. In that case users would get something like "Warning: hosted: pub.example.com will be interpreted as a package name. Try raising the minimum SDK version to 2.15 for pub to interpret it as a host" followed by `package pub.example.com not found on pub.dev`.

Reply:
Okay, let's throw a `FormatException` if:
 * Lower bound SDK constraint is less than 2.15 (short syntax not supported), and,
 * We see `hosted: <value>`, and,
 * `<value>` is not a valid package name without dots, ie. not matching `^[a-zA-Z_]+[a-zA-Z0-9_]*$`.

This will cause problems for packages containing `hosted: foo.bar.baz`, but such names can't be published to `pub.dev`, and if you are using a custom pub.dev server, then: You have to specify `hosted: {url: ..., name: ...}`, so you won't be able to use the `hosted: <name>`, unless you're overwriting `PUB_HOSTED_URL`.

So this would break users who:
 * custom pub servers, AND,
 * using package names containing dots (`.`), AND,
 * using the obscure syntax `hosted: <name>`.

All of those things happening at the same time seems unlikely. So this seems like a reasonable corner case to break.